docs/design/upgrade.md

```diff
@@ -0,0 +1,203 @@
+
+### <a name="change-api"></a>Change API
+A change is an operation that happens on string. It is an object that implements the following interface:
+
+```typescript
+interface Change {
+  // Apply the change.
+  apply(): Thenable<boolean>;
+
+  // The file this change should be applied to. Some changes might not apply to
+  // a file (maybe the config).
+  readonly path: string | null;
+
+  // The order this change should be applied. Normally the position inside the file.
+  // Changes are applied from the bottom of a file to the top.
+  readonly order: number | null;
+
+  // The description of this change. This will be outputted in a dry or verbose run.
+  readonly description: string;
+}
+```
+
+Some available example of a change implementation:
+
+```typescript
+// Will add text to the source code.
+class AppendChange implements Change {
+  constructor(file: string, pos: number, toAdd: string);
+}
+// Will remove text from the source code.
+class RemoveChange implements Change {
+  constructor(file: string, pos: number, toRemove: string);
+}
+// Will replace text from the source code.
+class ReplaceChange implements Change {
+  constructor(file: string, pos: number, oldText: string, newText: string);
+}
+// Will output a message for the user to fulfill.
+class MessageChange implements Change {
+  constructor(text: string);
+}
+```
+
+Some higher function changes are provided to allow an easier way to create changes that are harder to make:
+
+```typescript
+// Rename a class to another name. Need to pass a symbol as input, as just
+// a string is not enough.
+declare function renameClass(project: Angular2Project,
+                             from: ts.Symbol, to: string): Observable<Change>;
+// Upgrade a package in NPM.
+declare function upgradePackage(project: Angular2Project,
+                                name: string, to: string): Observable<Change>;
+// Change a configuration in angular-cli config.
+declare function configSet(project: Angular2Project, key: string,
+                           value: any): Observable<Change>;
+// Remove a file.
+declare function deleteFile(project: Angular2Project,
+                            file: string): Observable<Change>;
+// Scaffold a new blueprints.
+declare function scaffold(project: Angular2Project, name: string,
+                          arguments: {[name: string]: string}): Observable<Change>;
+```
+
+Some higher level changes examples:
+
+```typescript
+// Add a new Import if the import isn't already in the file.
+class AddImportChange implements Change {
+  constructor(file: string, symbolName: string, importPath: string);
+}
+```
+
+# Upgrade Command
+A new command must be added to the CLI;
+
+    ng upgrade [options]
+    Upgrade the project to a newer version.
+
+    Options:
+      --package=<name>  The package to upgrade. By default, upgrade every packages
+                        that are outdated.
+      --to=<semver>     Move to a specific version of Angular. Only valid with
```

Comment:
A user should be able to specify `--to` without `--dry-run`, e.g. he might not want the latest version since other packages are incompatible, latest might be broken for stuff he needs, etc.